chia/types/generator_types.py

```diff
@@ -2,12 +2,38 @@
 
 from dataclasses import dataclass, field
 
+from chia_rs import Coin, G2Element
+from chia_rs.sized_ints import uint32
+
 from chia.types.blockchain_format.serialized_program import SerializedProgram
 from chia.util.streamable import Streamable, streamable
 
 
+# This holds what we need to pre validate a block generator
 @streamable
 @dataclass(frozen=True)
 class BlockGenerator(Streamable):
     program: SerializedProgram = field(default_factory=SerializedProgram.default)
+    # to run the block generator, we need the actual bytes of the previous
+    # generators it may reference. These are parameters passed in to the block
+    # generator
     generator_refs: list[bytes] = field(default_factory=list)
+
+
+# When we create a new block, this object holds the block generator and
+# additional information we need to create the UnfinishedBlock from it.
+# When creating a block, we still need to be able to run it, to compute its cost
+# and validate it. Therefore, this is a superset of the BlockGenerator class.
+@dataclass(frozen=True)
+class NewBlockGenerator(BlockGenerator):
```

Comment:
Is it correct to assume that:
1. `Option[NewBlockGenerator]` is analogous to `Option[Tuple[BlockGenerator, G2Element, list[Coin], list[Coin]]]`
2. `Option[NewBlockGenerator]` is analogous to `Tuple[Option[BlockGenerator], G2Element, Option[list[Coin]], Option[list[Coin]]]` and the various other combinations, and all of the fields were previously either "default value" or filled with actual data?

If so, this seems like a massive improvement

Reply:
I would say it's closer to (1). if we don't have a generator, the signature is always default constructed and additions and removals are empty lists.